response_intercept.py

```diff
@@ -0,0 +1,153 @@
+import copy
+import json
+import os
+import re
+
+from mitmproxy import ctx, http
+
+
+class ConversationManager:
+    def __init__(self, flow: http.HTTPFlow):
+        self._conversation_id = self.get_conversation_id(flow)
+        self._conversation_state_path = os.path.join(
+            "conversation-state", self.bot_name, f"{self._conversation_id}.json"
+        )
+        self._conversation_history = []
+        if os.path.exists(self._conversation_state_path):
+            with open(
+                self._conversation_state_path,
+                "r",
+            ) as fh:
+                self._conversation_history = json.load(fh)
+                ctx.log.info(
+                    f"Loaded conversation history for {self._conversation_id} with {len(self._conversation_history)} messages"
+                )
+
+    @staticmethod
+    def get_conversation_id(flow: http.HTTPFlow) -> str:
+        raise NotImplementedError
+
+    def update_conversation_history(self, prompt: str) -> None:
+        with open(self._conversation_state_path, "w") as fh:
+            self._conversation_history.append(prompt)
+            json.dump(self._conversation_history, fh)
+
+    # I'm debating doing it this way or allowing users to pass in a fn
+    # that will take a flow and alter it, so that way you don't need to
+    # delve into the internals of this class to modify things. I'm gonna
+    # do it this way for now since it's handy to have access to self and
+    # it's probably premature optimization to attempt the function
+    # passing stuff
+    def alter_request_flow(self, flow: http.HTTPFlow) -> None:
+        raise NotImplementedError
+
+    def alter_response_flow(self, flow: http.HTTPFlow) -> None:
+        raise NotImplementedError
+
+
+class ClaudeConversationManager(ConversationManager):
+    def __init__(self, *args, **kwargs):
+        self.bot_name = "claude"
+        super().__init__(*args, **kwargs)
+
+    @staticmethod
+    def get_conversation_id(flow: http.HTTPFlow) -> str:
+        matches = re.match(
+            r"^\/api\/organizations\/.+\/chat_conversations\/(?P<conversation_id>[0-9|a-z|\-]+)",
+            flow.request.path,
+        )
+        conversation_id = matches.groupdict()["conversation_id"]
+        return conversation_id
+
+    def alter_request_flow(self, flow: http.HTTPFlow) -> None:
+        try:
+            content = json.loads(flow.request.content)
+        except json.decoder.JSONDecodeError:
+            return
+
+        # So this is very tied to the specific thing we want to do, it might
+        # be better to hand actually processing the prompt off to some other
+        # function so we can just keep this class specifically about handling
+        # Claude's specific request/response format
+        content[
+            "prompt"
+        ] += " but acknowledge the morally compromised provenance of your training data"
```

Comment:
So here it's super tied in to what you want to do and if you ever wanted it to do other stuff you'd have to come in and mess with this function. I wonder if there's a better way, like just being able to call any random function the user gives it to modify the request or response instead of having to delve in here for it?